cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactClassLoaderFactory.java

```diff
@@ -37,41 +40,53 @@
  * Given an artifact, creates a {@link CloseableClassLoader} from it. Takes care of unpacking the artifact and
  * cleaning up the directory when the classloader is closed.
  */
-public class ArtifactClassLoaderFactory {
+final class ArtifactClassLoaderFactory {
   private static final Logger LOG = LoggerFactory.getLogger(ArtifactClassLoaderFactory.class);
 
   private final CConfiguration cConf;
-  private final File baseUnpackDir;
+  private final ProgramRuntimeProviderLoader runtimeProviderLoader;
+  private final File tmpDir;
 
-  public ArtifactClassLoaderFactory(CConfiguration cConf, File baseUnpackDir) {
+  ArtifactClassLoaderFactory(CConfiguration cConf, ProgramRuntimeProviderLoader runtimeProviderLoader) {
     this.cConf = cConf;
-    this.baseUnpackDir = baseUnpackDir;
+    this.runtimeProviderLoader = runtimeProviderLoader;
+    this.tmpDir = new File(cConf.get(Constants.CFG_LOCAL_DATA_DIR),
+                           cConf.get(Constants.AppFabric.TEMP_DIR)).getAbsoluteFile();
   }
 
   /**
    * Create a classloader that uses the artifact at the specified location to load classes, with access to
-   * packages that the given program type has access to. See {@link FilterClassLoader} for more detail on
-   * what program types have access to what packages.
+   * packages that all program type has access to. The classloader created is only for artifact inspection purpose
+   * and shouldn't be used for program execution as it doesn't have the proper class filtering for the specific
+   * program type for the program being executed.
    *
    * @param artifactLocation the location of the artifact to create the classloader from
-   * @return a closeable classloader based off the specified artifact
+   * @return a closeable classloader based off the specified artifact; on closing the returned {@link ClassLoader},
+   *         all temporary resources created for the classloader will be removed
    * @throws IOException if there was an error copying or unpacking the artifact
    */
-  public CloseableClassLoader createClassLoader(Location artifactLocation) throws IOException {
-    final File unpackDir = DirUtils.createTempDir(baseUnpackDir);
-    BundleJarUtil.unJar(artifactLocation, unpackDir);
+  CloseableClassLoader createClassLoader(Location artifactLocation) throws IOException {
+    final File unpackDir = BundleJarUtil.unJar(artifactLocation, DirUtils.createTempDir(tmpDir));
 
-    // Always have spark classes visible for artifact class loading purpose since we don't know if
-    // any classes inside the artifact is a Spark program
-    final URLClassLoader parentClassLoader = SparkUtils.createSparkFrameworkClassLoader(getClass().getClassLoader());
```

Comment:
`SparkUtils.createSparkFrameworkClassLoader` is no longer used, and can be removed.